src/udev/udevadm-info.c

```diff
@@ -254,6 +254,53 @@ static void cleanup_dir(DIR *dir, mode_t mask, int depth) {
         }
 }
 
+/*
+ * Assume that dir is a directory with file names matching udev data base
+ * entries for devices in /run/udev/data (such as "b8:16"), and removes
+ * all files except those that haven't been deleted in /run/udev/data
+ * (i.e. they were skipped during db cleanup because of the db_persist flag).
+ * Returns true if the directory is empty after cleanup.
+ */
+static bool cleanup_dir_after_db_cleanup(DIR *dir, DIR *datadir) {
+        unsigned int kept = 0;
+
+        FOREACH_DIRENT_ALL(dent, dir, break) {
+                struct stat data_stats, link_stats;
+
+                if (dot_or_dot_dot(dent->d_name))
+                        continue;
+                if (fstatat(dirfd(dir), dent->d_name, &link_stats, AT_SYMLINK_NOFOLLOW) != 0)
+                        continue;
+
+                if (fstatat(dirfd(datadir), dent->d_name, &data_stats, 0) != 0)
+                        (void) unlinkat(dirfd(dir), dent->d_name,
+                                        S_ISDIR(link_stats.st_mode) ? AT_REMOVEDIR : 0);
+                else
+                        kept++;
```

Comment:
I think the counter should be also increased in the above two `continue` cases.

Reply:
certainly not for `.` or `..`, or am I missing something? In the case when `fstatat()` errors out, I suppose `keep` should only be incremented for certain error codes. Not for `ENOENT`, for example.

Reply:
resolved with b148f0cc26